File: apps/web/src/api/intelligence.ts
/**
 * AI Intelligence API client — LLM-powered smart features.
 */
import { apiJson } from "./client";

// ── Smart synthesis params ──────────────────────────────────────

export interface SmartSynthParamsResult {
  emotion: string;
  emotion_label: string;
  emotion_strength: number;
  speed_factor: number;
  temperature: number;
  pitch_factor: number;
  reasoning: string;
}

export interface SmartSynthParamsResponse {
  result: SmartSynthParamsResult;
  mode: "llm" | "fallback";
}

export async function recommendSynthParams(params: {
  text: string;
  character_hint?: string;
  context_hint?: string;
}): Promise<SmartSynthParamsResponse> {
  return apiJson<SmartSynthParamsResponse>("/api/v1/intelligence/synth-params", {
    method: "POST",
    body: JSON.stringify(params),
  });
}

// ── Smart voice matching ───────────────────────────────────────

export interface SmartVoiceMatchItem {
  voice_id: string;
  score: number;
  reason: string;
}

export interface SmartVoiceMatchResponse {
  matches: SmartVoiceMatchItem[];
  mode: "llm" | "fallback";
}

export async function matchVoice(params: {
  character_description: string;
  available_voices: Array<Record<string, unknown>>;
}): Promise<SmartVoiceMatchResponse> {
  return apiJson<SmartVoiceMatchResponse>("/api/v1/intelligence/match-voice", {
    method: "POST",
    body: JSON.stringify(params),
  });
}

// ── Smart emotion ──────────────────────────────────────────────

export interface SmartEmotionResult {
  emotion: string;
  emotion_label: string;
  strength: number;
  text_preview: string;
  mode: "llm" | "keyword";
}

export async function analyzeEmotionSmart(
  text: string,
  useLLM = true,
): Promise<SmartEmotionResult> {
  const params = new URLSearchParams({ text, use_llm: String(useLLM) });
  return apiJson<SmartEmotionResult>(`/api/v1/intelligence/emotion?${params}`, {
    method: "POST",
  });
}

// ── Content moderation ─────────────────────────────────────────

export interface SmartModerateResult {
  passed: boolean;
  risk_level: "low" | "medium" | "high";
  flags: string[];
  reason: string;
}

export interface SmartModerateResponse {
  result: SmartModerateResult;
  mode: "llm" | "rule";
}

export async function moderateContent(params: {
  text: string;
  context?: "post" | "message" | "profile" | "voice_description";
}): Promise<SmartModerateResponse> {
  return apiJson<SmartModerateResponse>("/api/v1/intelligence/moderate", {
    method: "POST",
    body: JSON.stringify(params),
  });
}

// ── Voice description generation ───────────────────────────────

export interface GeneratedVoiceDescription {
  title: string;
  description: string;
  tags: string[];
  suitable_for: string[];
  mode: "llm" | "fallback";
}

export async function generateVoiceDescription(params: {
  voice_name: string;
  tags?: string[];
  sample_text?: string;
}): Promise<GeneratedVoiceDescription> {
  const searchParams = new URLSearchParams({ voice_name: params.voice_name });
  if (params.tags?.length) {
    searchParams.set("tags", params.tags.join(","));
  }
  if (params.sample_text) {
    searchParams.set("sample_text", params.sample_text);
  }
  return apiJson<GeneratedVoiceDescription>(
    `/api/v1/intelligence/voice-description?${searchParams}`,
    { method: "POST" },
  );
}

// ── Script polish ────────────────────────────────────────────────

export interface ScriptPolishResponse {
  polished_text: string;
  changes_summary: string;
  character_names: string[];
  line_count: number;
  mode: "llm" | "fallback";
}

export async function polishScript(params: {
  text: string;
  polish_scope?: "full" | "grammar" | "names" | "narration";
}): Promise<ScriptPolishResponse> {
  return apiJson<ScriptPolishResponse>("/api/v1/intelligence/polish-script", {
    method: "POST",
    body: JSON.stringify({
      text: params.text,
      polish_scope: params.polish_scope ?? "full",
    }),
  });
}

File: apps/api/routes/intelligence.py
"""Intelligence API routes — LLM-powered smart features."""

from __future__ import annotations

from uuid import UUID

from apps.api.deps import get_current_user_id
from apps.api.exceptions import raise_domain_http
from domains.intelligence.schemas import (
    ScriptPolishRequest,
    ScriptPolishResponse,
    SmartModerateRequest,
    SmartModerateResponse,
    SmartSynthParamsRequest,
    SmartSynthParamsResponse,
    SmartVoiceMatchRequest,
    SmartVoiceMatchResponse,
)
from domains.intelligence.service import IntelligenceService, IntelligenceServiceError
from fastapi import APIRouter, Depends, Query

router = APIRouter()
_service = IntelligenceService()


@router.post("/intelligence/synth-params", response_model=SmartSynthParamsResponse)
def recommend_synth_params(
    body: SmartSynthParamsRequest,
    _: UUID = Depends(get_current_user_id),
) -> SmartSynthParamsResponse:
    """AI 智能推荐合成参数（情感、语速、温度、音高）。"""
    try:
        return _service.recommend_synth_params(
            text=body.text,
            character_hint=body.character_hint,
            context_hint=body.context_hint,
        )
    except IntelligenceServiceError as exc:
        raise_domain_http(exc)


@router.post("/intelligence/match-voice", response_model=SmartVoiceMatchResponse)
def match_voice(
    body: SmartVoiceMatchRequest,
    _: UUID = Depends(get_current_user_id),
) -> SmartVoiceMatchResponse:
    """AI 智能音色匹配 — 根据角色描述推荐最合适的音色。"""
    try:
        return _service.match_voice(body)
    except IntelligenceServiceError as exc:
        raise_domain_http(exc)


@router.post("/intelligence/emotion", response_model=dict)
def analyze_emotion_smart(
    text: str = Query(min_length=1, max_length=5000, description="待分析文本"),
    use_llm: bool = Query(default=True, description="是否使用 LLM 增强分析"),
    _: UUID = Depends(get_current_user_id),
) -> dict:
    """AI 增强情感分析 — LLM 语义理解 + 关键词兜底。"""
    try:
        return _service.analyze_emotion_smart(text, use_llm=use_llm)
    except IntelligenceServiceError as exc:
        raise_domain_http(exc)


@router.post("/intelligence/moderate", response_model=SmartModerateResponse)
def moderate_content(
    body: SmartModerateRequest,
    _: UUID = Depends(get_current_user_id),
) -> SmartModerateResponse:
    """AI 内容审核 — 智能检测违规内容。"""
    try:
        return _service.moderate_content(body)
    except IntelligenceServiceError as exc:
        raise_domain_http(exc)


@router.post("/intelligence/voice-description", response_model=dict)
def generate_voice_description(
    voice_name: str = Query(min_length=1, max_length=64, description="音色名称"),
    tags: str | None = Query(default=None, description="现有标签，逗号分隔"),
    sample_text: str | None = Query(default=None, max_length=500, description="样音文本"),
    _: UUID = Depends(get_current_user_id),
) -> dict:
    """AI 智能生成音色馆文案（标题、描述、标签、适用场景）。"""
    try:
        tag_list = [t.strip() for t in tags.split(",") if t.strip()] if tags else None
        return _service.generate_voice_description(
            voice_name=voice_name,
            tags=tag_list,
            sample_text=sample_text,
        )
    except IntelligenceServiceError as exc:
        raise_domain_http(exc)


@router.post("/intelligence/polish-script", response_model=ScriptPolishResponse)
def polish_script(
    body: ScriptPolishRequest,
    _: UUID = Depends(get_current_user_id),
) -> ScriptPolishResponse:
    """AI 剧本智能润色 — 修正语法、统一角色名、增强旁白。"""
    try:
        return _service.polish_script(body)
    except IntelligenceServiceError as exc:
        raise_domain_http(exc)
